### PR TITLE
chore(engine): refactor pipeline outcome processing

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Refactor and simplify the processing of pipeline outcome.

Remove the anti-pattern of returning `Option<Result<(), _>>` where the only use case for returning `Some(Ok(()))` was to indicate that the pipeline has reached the max block.

Gets rid of `match result` statements and uses residual instead.

### CODE REVIEW

File: crates/consensus/beacon/src/engine/mod.rs [hunk 1]
```diff
@@ -1564,168 +1554,144 @@ where
...
             EngineSyncEvent::PipelineStarted(target) => {
                 trace!(target: "consensus::engine", ?target, continuous = target.is_none(), "Started the pipeline");
                 self.metrics.pipeline_runs.increment(1);
                 self.sync_state_updater.update_sync_state(SyncState::Syncing);
+                SyncEventOutcome::Processed
+            }
+            EngineSyncEvent::PipelineFinished { result, reached_max_block } => {
+                trace!(target: "consensus::engine", ?result, ?reached_max_block, "Pipeline finished");
+                // Any pipeline error at this point is fatal.
+                let ctrl = result?;
```
Comment (rkrasiuk) [inline crates/consensus/beacon/src/engine/mod.rs:1572]: any pipeline error is considered fatal, same as before

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -1564,168 +1554,144 @@ where
...
-                    )
-                    .is_none()
-                {
-                    // get the block number of the finalized block, if we have it
-                    let newest_finalized = self
-                        .forkchoice_state_tracker
-                        .sync_target_state()
-                        .map(|s| s.finalized_block_hash)
-                        .and_then(|h| self.blockchain.buffered_header_by_hash(h))
-                        .map(|header| header.number);
```
Comment (rkrasiuk) [inline crates/consensus/beacon/src/engine/mod.rs:1680]: this unnecessarily retrieved optional `self.forkchoice_state_tracker.sync_target_state()`, but we already retrieved and ensured that it's some up above

File: crates/consensus/beacon/src/engine/mod.rs [hunk 3]
```diff
@@ -1564,168 +1554,144 @@ where
...
-                    // pipeline target
-                    let pipeline_target = if let (Some(progress), Some(finalized_number)) =
-                        (ctrl.block_number(), newest_finalized)
-                    {
-                        // Determines whether or not we should run the pipeline again, in case the
-                        // new gap is large enough to warrant running the pipeline.
-                        self.can_pipeline_sync_to_finalized(progress, finalized_number, None)
-                    } else {
-                        None
-                    };
```
Comment (rkrasiuk) [inline crates/consensus/beacon/src/engine/mod.rs:1695]: `if let (Some(), Some()) { /* invoke fn that returns an option */ } else { None }` is simply refactored to `.zip().and_then(...)`

**Comment by rkrasiuk** [comment]
Preemptive hive run, because why not
https://github.com/paradigmxyz/reth/actions/runs/8718483460

**Review by mattsse — APPROVED** [review]
all of these look okay.

pending hive

**Comment by rkrasiuk** [comment]
hive results are unchanged

**Pull request merged**